### PR TITLE
TLS authentication implementation

### PR DESCRIPTION
**Pull request by RahimovIR**

### CODE REVIEW

File: src/Pulsar.Client/Internal/ConnectionPool.fs [hunk 1]
```diff
@@ -60,7 +61,39 @@ type internal ConnectionPool (config: PulsarClientConfiguration) =
...
+                        let status = lastChainElm.ChainElementStatus.[0].Status
+                        if status = X509ChainStatusFlags.UntrustedRoot then
+                            if System.Linq.Enumerable.SequenceEqual(lastChainElm.Certificate.RawData, ca.RawData) then
+                                Log.Logger.LogDebug("Used TlsTrustCertificate")
+                                true
+                            else
+                                Log.Logger.LogError("TlsTrustCertificate no equal with root certificate")
+                                false
+                        else
+                            Log.Logger.LogError("Root certificate status {} no equal UntrustedRoot", status)
```
Comment (Lanayx) [inline src/Pulsar.Client/Internal/ConnectionPool.fs]: Should be {0}

File: src/Pulsar.Client/Internal/ConnectionPool.fs [hunk 2]
```diff
@@ -74,18 +107,18 @@ type internal ConnectionPool (config: PulsarClientConfiguration) =
...
-                    use cert2 = new X509Certificate2(cert)
-                    chain.Build(cert2)
+            let result = config.TlsAllowInsecureConnection || CheckRemoteCertWithTrustCertificate()
             let log = if result then Log.Logger.LogWarning else Log.Logger.LogError
             log("RemoteCertificateChainErrors")
             result
+        | error when error = SslPolicyErrors.RemoteCertificateChainErrors + SslPolicyErrors.RemoteCertificateNameMismatch ->
+            let result =
+                not config.TlsHostnameVerificationEnable
+                &&
```
Comment (Lanayx) [inline src/Pulsar.Client/Internal/ConnectionPool.fs]: Can you please move it to the upper line

File: src/Pulsar.Client/Internal/ConnectionPool.fs [hunk 3]
```diff
@@ -60,7 +61,39 @@ type internal ConnectionPool (config: PulsarClientConfiguration) =
...
+                                Log.Logger.LogDebug("Used TlsTrustCertificate")
+                                true
+                            else
+                                Log.Logger.LogError("TlsTrustCertificate no equal with root certificate")
+                                false
+                        else
+                            Log.Logger.LogError("Root certificate status {} no equal UntrustedRoot", status)
+                            false
+                    else 
+                        Log.Logger.LogError("Root certificate status count > 1, status: {}", lastChainElm.ChainElementStatus)
```
Comment (Lanayx) [inline src/Pulsar.Client/Internal/ConnectionPool.fs]: Should be {0}

File: src/Pulsar.Client/Internal/ConnectionPool.fs [hunk 4]
```diff
@@ -60,7 +61,39 @@ type internal ConnectionPool (config: PulsarClientConfiguration) =
...
+                        false
+                else
+                    Log.Logger.LogError("Builds an X.509 chain faild, status: {}", chain.ChainStatus)
```
Comment (Lanayx) [inline src/Pulsar.Client/Internal/ConnectionPool.fs]: Should be {0}

File: examples/FsharpExamples/Program.fs [hunk 5]
```diff
@@ -19,6 +19,7 @@ let main _ =
     PulsarClient.Logger <- loggerFactory.CreateLogger("PulsarLogger") 
 
     runSimple().Wait()
+    runTlsAuthentication().Wait()
```
Comment (Lanayx) [inline examples/FsharpExamples/Program.fs]: Please comment this line as well as lines below, so only Simple will be run by default

File: examples/FsharpExamples/TlsAuthentication.fs [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+open Pulsar.Client.Api
+open FSharp.Control.Tasks.V2.ContextInsensitive
+open Pulsar.Client.Common
+open Pulsar.Client.Api
+open System.Text
+open System.Security.Cryptography.X509Certificates
+
+let runTlsAuthentication () =
+
+    let serviceUrl = "pulsar://my-pulsar-cluster:30002"
```
Comment (Lanayx) [inline examples/FsharpExamples/TlsAuthentication.fs]: Please change to "pulsar+ssl//..." here

File: src/Pulsar.Client/Auth/AuthenticationDataTls.fs [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+open Pulsar.Client.Api
+
+type internal AuthenticationDataTls(certFile: FileInfo) =
+    inherit AuthenticationDataProvider()
+    
+    let createCertCollection() =
+        use cert = new X509Certificate2(certFile.FullName)
+        if cert.HasPrivateKey then
+            let result  = X509CertificateCollection()
+            result.Add(new X509Certificate(certFile.FullName)) |> ignore
```
Comment (Lanayx) [inline src/Pulsar.Client/Auth/AuthenticationDataTls.fs]: why do we create X509Certificate twice?

File: examples/FsharpExamples/FsharpExamples.fsproj [hunk 8]
```diff
@@ -10,6 +10,7 @@
     <Compile Include="Reader.fs" />
     <Compile Include="Simple.fs" />
     <Compile Include="RealWorld.fs" />
+    <Compile Include="TlsAuthentication.fs" />
```
Comment (Lanayx) [inline examples/FsharpExamples/FsharpExamples.fsproj:13]: Can you please add C# examples as well.

File: src/Pulsar.Client/Auth/AuthenticationDataTls.fs (hunk 7 above)
Comment (RahimovIR) [reply src/Pulsar.Client/Auth/AuthenticationDataTls.fs]: X509Certificate must be put in the collection, and using X509Certificate2 we check for the presence of the private key

Comment (Lanayx) [reply src/Pulsar.Client/Auth/AuthenticationDataTls.fs]: X509Certificate2 can be easily casted to X509Certificate

File: src/Pulsar.Client/Auth/AuthenticationDataTls.fs [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+            result
+        else
+            X509CertificateCollection()
```
Comment (Lanayx) [inline src/Pulsar.Client/Auth/AuthenticationDataTls.fs]: Why do we return empty collection for this case? It seems that throwing an exception "Certificate doesn't contain a private key" is a more user-friendly way to handle it

File: src/Pulsar.Client/Internal/ConnectionPool.fs (hunk 1 above)
Comment (RahimovIR) [reply src/Pulsar.Client/Internal/ConnectionPool.fs]: Fixed

File: src/Pulsar.Client/Internal/ConnectionPool.fs (hunk 2 above)
Comment (RahimovIR) [reply src/Pulsar.Client/Internal/ConnectionPool.fs]: Fixed

File: src/Pulsar.Client/Internal/ConnectionPool.fs (hunk 3 above)
Comment (RahimovIR) [reply src/Pulsar.Client/Internal/ConnectionPool.fs]: Fixed

File: src/Pulsar.Client/Internal/ConnectionPool.fs (hunk 4 above)
Comment (RahimovIR) [reply src/Pulsar.Client/Internal/ConnectionPool.fs]: Fixed

File: examples/FsharpExamples/TlsAuthentication.fs (hunk 6 above)
Comment (RahimovIR) [reply examples/FsharpExamples/TlsAuthentication.fs]: Fixed

File: src/Pulsar.Client/Auth/AuthenticationDataTls.fs (hunk 7 above)
Comment (RahimovIR) [reply src/Pulsar.Client/Auth/AuthenticationDataTls.fs]: But after casting, authorization doesn't work.

File: src/Pulsar.Client/Auth/AuthenticationDataTls.fs (hunk 9 above)
Comment (RahimovIR) [reply src/Pulsar.Client/Auth/AuthenticationDataTls.fs]: Fixed

**Pull request merged**